### PR TITLE
[SPIR-V] Stop re-writing SPIR-V type map

### PR DESCRIPTION
**Pull request by MrSidims**
There is a map used during forward translation. It links LLVM type
and SPIR-V type, that is being generated, LLVM type here is used as
a key. Due to a bug in type mapping (information is below), it's
possible that for a single pointer type in LLVM, the translator
will create 2 physically-indentical SPIR-V pointer types and with
previous implementation of mapType function the map would be changed
during the second type insertion. This patch prevents it, forcing
to reuse the type that was created first, though we would still have
an unused duplication of this type in SPIR-V module.

Please consider following LLVM IR sample:
%struct._ZTS4Args.Args = type { %struct._ZTS6Object.Object addrspace(4)* }
%struct._ZTS6Object.Object = type { i32 (%struct._ZTS6Object.Object addrspace(4)*, i32)* }

Here we have 2 LLVM type declarations. The translator will recursively
go through structure members and then through element types of the
pointer types, creating the appropriate SPIR-V types and mapping them
on LLVM types. Due to a recursive nature of this algorithm, here we will
have '%struct._ZTS6Object.Object addrspace(4)*' be processed twice with
2 SPIR-V type declaration instruction being created.

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

### CODE REVIEW

**Comment by MrSidims** [comment]
I would prefer to push this patch ASAP not waiting for Khronos review and a pull-down.

**Review by AlexeySotkin — APPROVED** [review]

**Comment by bader** [comment]
> I would prefer to push this patch ASAP not waiting for Khronos review and a pull-down.

Why?

**Comment by MrSidims** [comment]
> > I would prefer to push this patch ASAP not waiting for Khronos review and a pull-down.
> 
> Why?

There is an assertion firing because of this bug (store instruction validation) and a customer, that uses intel/llvm build wants it be fixed ASAP.

File: llvm-spirv/test/pointer_type_mapping.ll [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+; ModuleID = 'sycl_test.bc'
+source_filename = "sycl_test.cpp"
+target datalayout = "e-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024-n8:16:32:64"
+target triple = "spir64-unknown-unknown-sycldevice"
+
+%struct._ZTS4Args.Args = type { %struct._ZTS6Object.Object addrspace(4)* }
+%struct._ZTS6Object.Object = type { i32 (%struct._ZTS6Object.Object addrspace(4)*, i32)* }
+
+; Function Attrs: convergent norecurse nounwind mustprogress
+define dso_local spir_func i32 @_Z9somefunc0P4Args(%struct._ZTS4Args.Args addrspace(4)* %args) #0 {
```
Comment (bader) [inline llvm-spirv/test/pointer_type_mapping.ll:22]: Please, reduce the test case.
```suggestion
define dso_local spir_func i32 @_Z9somefunc0P4Args(%struct._ZTS4Args.Args addrspace(4)* %a, %struct._ZTS6Object.Object addrspace(4)* %b) {
}
```
Replace everything below this line with
```llvm
!opencl.spir.version = !{!0}
!spirv.Source = !{!1}
!0 = !{i32 1, i32 2}
!1 = !{i32 4, i32 100000}
```

This should be enough to reproduce the problem with types translation. Right?

**Review by bader — CHANGES_REQUESTED** [review]

File: llvm-spirv/test/pointer_type_mapping.ll (hunk 1 above)
Comment (MrSidims) [reply llvm-spirv/test/pointer_type_mapping.ll]: The test won't check if the assertion comes from validation of store instruction, but I believe you are right. Applied.

**Review by bader — APPROVED** [review]

**Review by AlexeySachkov — APPROVED** [review]
@MrSidims, could you please report CI failure - it is not the first time I see that something is wrong on some download steps.

The patch looks good to me - we have https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1048 submitted to track the main problem and this patch seems to be a small, but reasonable step towards proper solution

**Comment by MrSidims** [comment]
> @MrSidims, could you please report CI failure - it is not the first time I see that something is wrong on some download steps.
> 
> The patch looks good to me - we have [KhronosGroup/SPIRV-LLVM-Translator#1048](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1048) submitted to track the main problem and this patch seems to be a small, but reasonable step towards proper solution

I reported it. Though, while I'm trying to get access to the logs I'm getting a timeout error.

**Pull request merged**